### PR TITLE
Doi create operation

### PR DESCRIPTION
**Pull request by mike-podolskiy90**
First version of the DataCite API client.
Some stuff should be added:
- Jenkins file (enable pipeline)
- sonar and jacoco
- unit/integration tests should be improved (additional and edge cases are needed)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,3 +10,61 @@ The implementation of this library will depend on the following commonly used li
  - [SLF4J](https://www.slf4j.org/) as logging facade.
```
Comment (fmendezh) [inline README.md:20]: Please add README files to submodules

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
+package org.gbif.datacite.rest.configuration;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+public class ClientConfiguration implements Serializable {
```
Comment (fmendezh) [inline datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:11]: Add Javadoc

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    private static final long serialVersionUID = 7778374827110638452L;
+
+    private final String baseApiUrl;
+
+    private final Long timeOut;
```
Comment (fmendezh) [inline datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:17]: default?

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    private final Long fileCacheMaxSizeMb;
```
Comment (fmendezh) [inline datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:19]: default?

File: datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import retrofit2.Retrofit;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+class DataCiteRetrofitServiceTest {
```
Comment (fmendezh) [inline datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java:30]: javadoc

File: datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    private static MockWebServer server;
+    private static DataCiteRetrofitService service;
+    private static ResourceConverter converter;
```
Comment (fmendezh) [inline datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java]: move it to a local variable

File: datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        // given
+        String response = IOUtils.getResourceAsString("draftDoiResponse.json");
+
+        server.enqueue(new MockResponse()
+                .setResponseCode(200)
+                .setBody(response)
+        );
+
+        // when
+        Response<JSONAPIDocument<Datacite42Schema>> actual = service.get("10.21373/110").execute();
```
Comment (fmendezh) [inline datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java:80]: move "10.21373/110" to constant ?

File: datacite-rest-client/src/test/java/org/gbif/datacite/util/IOUtils.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+/**
+ * Provides utility methods for IO related actions.
+ */
+public class IOUtils {
```
Comment (fmendezh) [inline datacite-rest-client/src/test/java/org/gbif/datacite/util/IOUtils.java:11]: add private constructor

File: datacite-rest-client/src/test/java/org/gbif/datacite/util/IOUtils.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * For getting file content by name as a string.
+     *
+     * @param name file name
+     * @return character data (String)
+     * @throws IOException in case of some input/output problems
+     */
+    public static String getResourceAsString(String name) throws IOException {
+        InputStream input = IOUtils.class.getClassLoader().getResourceAsStream(name);
```
Comment (fmendezh) [inline datacite-rest-client/src/test/java/org/gbif/datacite/util/IOUtils.java:21]: maybe something like this does the job new String (Files.readAllBytes(name File(IOUtils.class.getClassLoader().getResource(name).toURI())), StandardCharsets.UTF_8); ?

File: datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceIT.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package org.gbif.datacite.rest.client.retrofit;
+
+import cucumber.api.CucumberOptions;
+import cucumber.api.junit.Cucumber;
+import org.junit.runner.RunWith;
+
+@RunWith(Cucumber.class)
```
Comment (fmendezh) [inline datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceIT.java:12]: javadoc

File: datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+import org.gbif.datacite.model.json.Datacite42Schema;
+import org.gbif.datacite.rest.configuration.ClientConfiguration;
+import org.gbif.datacite.rest.retrofit.RetrofitClientFactory;
+import retrofit2.Response;
+
+import java.io.IOException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DataCiteRetrofitServiceSteps {
```
Comment (fmendezh) [inline datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java:19]: javadoc

File: datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @When("^Perform a request to DataCite's PUT DOI$")
+    public void performUpdateDoi() throws IOException {
+        Datacite42Schema datacite42Schema = new Datacite42Schema();
+        datacite42Schema.setId(DOI_105);
+        datacite42Schema.setDoi(DOI_105);
+        datacite42Schema.setPublicationYear("2019");
+
+        JSONAPIDocument<Datacite42Schema> jsonapi = new JSONAPIDocument<>(datacite42Schema);
```
Comment (fmendezh) [inline datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java]: jsonapi  -> jsonApi

File: datacite-rest-client/pom.xml [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        <dependency>
+            <groupId>org.junit.vintage</groupId>
+            <artifactId>junit-vintage-engine</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.squareup.okhttp3</groupId>
+            <artifactId>mockwebserver</artifactId>
+        </dependency>
+
+        <!-- Integration testing (cucumber) -->
```
Comment (fmendezh) [inline datacite-rest-client/pom.xml:83]: mvn dependency:analyze reports this missing dependency io.cucumber:cucumber-core, is that correct?

File: datacite-jsonapi-annotator/pom.xml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>datacite-jsonapi-annotator</artifactId>
+
+    <name>DataCite JSON:API Annotator</name>
+    <description>DataCite JSON:API Annotator</description>
+
+    <dependencies>
```
Comment (fmendezh) [inline datacite-jsonapi-annotator/pom.xml:17]: missing dependency com.sun.codemodel:codemodel
are jackson core and annotations needed?

**Review by fmendezh — CHANGES_REQUESTED** [review]
Just minor changes, greate work!!

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java (hunk 2 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:11]: done

File: README.md (hunk 1 above)
Comment (mike-podolskiy90) [reply README.md:20]: done

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java (hunk 3 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:17]: ClientConfiguration can be constructed via all args constructor or builder. Builder provides default values.

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java (hunk 4 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:19]: the same as above

File: datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java (hunk 5 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java:30]: done

File: datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java (hunk 6 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java]: seems to be redundant. removed

File: datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceIT.java (hunk 10 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceIT.java:12]: done

File: datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java (hunk 11 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java:19]: done

File: datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java (hunk 12 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/it/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceSteps.java]: done

File: datacite-jsonapi-annotator/pom.xml (hunk 14 above)
Comment (mike-podolskiy90) [reply datacite-jsonapi-annotator/pom.xml:17]: mockito-inline is used. looks like false positive scanning result.
fix codemodel.
remove redundant jackson and others.

File: datacite-rest-client/pom.xml (hunk 13 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/pom.xml:83]: fixed

File: datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class JsonApiAnnotatorTest {
```
Comment (fmendezh) [inline datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java:24]: test cases names do not follow the camelCase format used for all the methods

Comment (fmendezh) [inline datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java:24]: Javadoc

**Review by fmendezh — CHANGES_REQUESTED** [review]
some comments/questions are still unanswered and there a couple of suggestions before merging this PR

File: datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java (hunk 15 above)
Comment (mike-podolskiy90) [reply datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java:24]: I usually use the test naming strategy testName_whenSomething_thenExpectSomethingElse. We used that on several past projects. 
If it's not ok I can change.
https://dzone.com/articles/7-popular-unit-test-naming

File: datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java (hunk 4 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/main/java/org/gbif/datacite/rest/configuration/ClientConfiguration.java:19]: It's provided here
`public static class Builder {
        ...
        private Long fileCacheMaxSizeMb = 64L;
...`

File: datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java (hunk 15 above)
Comment (mike-podolskiy90) [reply datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java:24]: in order to be consistent I will follow the naming strategy number 3 from the link above

Comment (mike-podolskiy90) [reply datacite-jsonapi-annotator/src/test/java/org/gbif/datacite/jsonapi/annotator/JsonApiAnnotatorTest.java:24]: done

File: datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java (hunk 6 above)
Comment (mike-podolskiy90) [reply datacite-rest-client/src/test/java/org/gbif/datacite/rest/client/retrofit/DataCiteRetrofitServiceTest.java]: done
